app/javascript/components/Editor.tsx

```diff
@@ -309,14 +251,96 @@ export function Editor({
     )
   }, [sendRequest, initialSubmission, updateSubmission])
 
-  const revertContent = useCallback(() => {
+  const revertToLastIteration = useCallback(() => {
     editorRef.current?.setFiles(initialFiles)
   }, [initialFiles])
 
   const toggleKeyboardShortcuts = useCallback(() => {
     editorRef.current?.openPalette()
   }, [editorRef])
 
+  const revertToExerciseStart = useCallback(() => {
+    if (!submission) {
+      return
+    }
+
+    revertDispatch({ type: RevertStatus.INITIALIZED })
+
+    sendRequest(submission.links.files, null, 'GET')
+      .then((json: any) => {
+        if (!json) {
+          return
+        }
+
+        const files = typecheck<File[]>(json, 'files')
+
+        editorRef.current?.setFiles(files)
+
+        revertDispatch({ type: RevertStatus.SUCCEEDED })
+      })
+      .catch((err) => {
+        if (err instanceof Error) {
+          revertDispatch({
+            type: RevertStatus.FAILED,
+            payload: {
+              apiError: {
+                type: 'unknown',
+                message: 'Unable to revert file, please try again.',
+              } as APIError,
+            },
+          })
+        }
+
+        if (err instanceof Response) {
+          err.json().then((res: any) => {
+            revertDispatch({
+              type: RevertStatus.FAILED,
+              payload: { apiError: res.error },
+            })
+          })
+        }
+      })
+  }, [revertDispatch, sendRequest, submission])
+
+  useEffect(() => {
+    switch (submissionStatus) {
+      case SubmissionStatus.CREATED:
+        setStatus(EditorStatus.SUBMISSION_CREATED)
+        break
+      case SubmissionStatus.CANCELLED:
+        setStatus(EditorStatus.SUBMISSION_CANCELLED)
+        break
+      case SubmissionStatus.CREATING:
+        setStatus(EditorStatus.CREATING_SUBMISSION)
+        break
+      case SubmissionStatus.CREATING_ITERATION:
+        setStatus(EditorStatus.CREATING_ITERATION)
+        break
+    }
+  }, [submissionStatus])
+
+  useEffect(() => {
+    switch (revertStatus) {
+      case RevertStatus.INITIALIZED:
+        setStatus(EditorStatus.REVERTING_TO_EXERCISE_START)
+        break
+      case RevertStatus.SUCCEEDED:
+        setStatus(EditorStatus.REVERTED)
+        break
+      case RevertStatus.FAILED:
+        setStatus(EditorStatus.REVERT_FAILED)
+        break
+    }
+  }, [revertStatus])
+
+  useEffect(() => {
```

Comment:
When an error comes in from either reverting and submitting, we set the API error.

Currently, we only have one place in the editor to show API errors. We can decide to have a separate place for each kind of error, but I decided to just put them in one place for now.